### PR TITLE
support AM/PM in date parsing/printing

### LINKED ISSUE #23870: Add `am/pm` to DateFormat
currently, DateFormat does not support `am/pm`. As this format is often used, it would be useful to be able to convert it to DateTime.

For Minutes after midday (e.g. 12:0-12:59 AM) and after midnight (0:01-0:59 PM), the am/pm-rule is a bit strange. Support in DateTime could help prevent bugs.

### PR DESCRIPTION
**Pull request by stevengj**
Support AM/PM specifier (`p` dateformat code, and `I` code for 12-hour clock) similar to strftime/strptime.  Closes #23870.

Follows the quasi-standard convention of 12am=midnight (of the following day), 12pm=noon (apparently followed by all other popular libraries and languages, as discussed in #23870).

```jl
julia> Time("8:10pm", dateformat"I:MMp")
20:10:00

julia> Dates.format(ans, dateformat"II:MM p")
"08:10 PM"
```

### CODE REVIEW

**Comment by galenlynch** [comment]
> 12am=midnight (of the following day)

I don't understand what you mean by "of the following day". Wouldn't "2019-03-22 12:00:00 AM" be the first second of 2019-03-22?

**Comment by stevengj** [comment]
> I don't understand what you mean by "of the following day". Wouldn't "2019-03-22 12:00:00 AM" be the first second of 2019-03-22?

Yes.

File: stdlib/Dates/src/io.jl [hunk 1]
```diff
@@ -161,6 +173,11 @@ for (tok, fn) in zip("uU", [monthabbr, monthname])
     end
 end
 
+function format(io, d::DatePart{'p'}, dt, locale)
+    ampm = hour(dt) < 12 ? "AM" : "PM" # fixme: locale-specific?
```
Comment (quinnj) [inline stdlib/Dates/src/io.jl]: As far as I can tell from some googling, there are only a handful of english-speaking countries that use AM/PM, so yeah, I'm not quite sure how critical it is to support locale here. For reference (https://www.timeanddate.com/time/am-and-pm.html)

**Review by quinnj — APPROVED** [review]

**Comment by stevengj** [comment]
Should be good to merge?

**Comment by omus** [comment]
It would been good to use `a` as the specifier for AM/PM which is used in the [Unicode Technical Standard #35](http://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table). I'll probably open a PR for this.

**Comment by stevengj** [comment]
@omus, the `p` comes from the POSIX `strftime`, which seems more consistent with our other specifiers.  e.g. we use `M` for minute and `m` for month, whereas UTS35 is the reverse.

**Comment by omus** [comment]
Completely reasonable. UTS35 has a standard on what using multiple specifiers in a row does which is a bit different from `strftime`. Following POSIX here is totally reasonable.

**Pull request merged**